### PR TITLE
feat: Create Launch Templates for each AZ - fixes #131

### LINKED ISSUE #131: Add support for Launch templates
This is a feature request.

It would be useful to support launch templates as they have few advantages
More details here: https://docs.aws.amazon.com/autoscaling/ec2/userguide/LaunchTemplates.html

This is an example I use:

```
  worker_groups_launch_template = [
    {
      name                                     = "spot-group-1"
      override_instance_types                  = local.spot_instance_types
      spot_instance_pools                      = local.spot_instance_pools
      asg_max_size                             = local.spot_asg_max_size
      asg_desired_capacity                     = local.spot_asg_desired_capacity
      public_ip                                = true
      default_cooldown                         = "300"
      asg_min_size                             = local.asg_min_size #set to 1 after autoscaler is in place
      autoscaling_enabled                      = "true"
      on_demand_base_capacity                  = "1"
      on_demand_allocation_strategy            = "prioritized"
      on_demand_percentage_above_base_capacity = "0"
      enabled_metrics                          = ["GroupMinSize", "GroupMaxSize", "GroupDesiredCapacity"]
      protect_from_scale_in                    = "true"
    },
  ]
```

### PR DESCRIPTION
**Pull request by patrickleet**
This sorta closes #131 - the original plan was to expose it directly, but it gets complicated and there are nuances around autoscaling so I figure this is what most people want although opinionated.

Perhaps a second PR could allow providing a completely custom launch template if someone needs that

### CODE REVIEW

File: modules/cluster/main.tf [hunk 1]
```diff
@@ -67,6 +67,34 @@ module "eks" {
   subnets         = (var.cluster_in_private_subnet ? module.vpc.private_subnets : module.vpc.public_subnets)
   vpc_id          = module.vpc.vpc_id
   enable_irsa     = true
+
+  worker_groups_launch_template = var.enable_worker_groups_launch_template ? [
+    for subnet in module.vpc.public_subnets :
```
Comment (ankitm123) [inline modules/cluster/main.tf:72]: Proabably have to do something similar for private subnets, ignore for this PR ...

File: README.md [hunk 2]
```diff
@@ -374,6 +380,55 @@ You can save up to 90% of cost when you use Spot Instances. You just need to mak
...
+Using only one often results in Spot instances not being able to be provisioned, and this greatly reduces the occurence of this happening, as well as allowing for lower spot prices.
+
+#### Enabling Worker Group Launch Templates
+
+To use the Worker Group Launch Template, set the variable `enable_worker_groups_launch_template` to `true`, and define an array of instance types allowed.
+
+When using autoscaling with Launch Templates per AZ, the min and max number of nodes is per zone.
+These values can be adjusted by using the variables `lt_desired_nodes_per_subnet`, `lt_min_nodes_per_subnet`, and `lt_max_nodes_per_subnet`
+
+```terraform
```
Comment (ankitm123) [inline README.md:402]: Move this from the markdown to the examples. I am working on a PR to add tests, so planning to just test everything in the examples folder.

File: variables.tf [hunk 3]
```diff
@@ -322,3 +322,33 @@ variable "ignoreLoadBalancer" {
...
+  description = "The minimum number of worker nodes in each Subnet (AZ) if using Launch Templates"
+  type        = number
+  default     = 1
+}
+
+variable "lt_max_nodes_per_subnet" {
+  description = "The maximum number of worker nodes in each Subnet (AZ) if using Launch Templates"
+  type        = number
+  default     = 2
+}
```
Comment (ankitm123) [inline variables.tf:354]: Add new line

**Comment by ankitm123** [comment]
Overall looks nice :+1: , left a few comments @patrickleet :)

**Comment by patrickleet** [comment]
the alternative of using the legacy worker groups literally doesn't work with autoscaler cause it uses all nodes from same AZ and eventually things can't be scheduled because of volume affinity, so I think it's a pretty good improvement :) plus you can get cheaper nodes scheduled more consistently.

I'll address these changes and rebase now!

**Pull request merged**